rclcpp/include/rclcpp/strategies/message_pool_memory_strategy.hpp

```diff
@@ -87,24 +99,68 @@ class MessagePoolMemoryStrategy
    */
   void return_message(std::shared_ptr<MessageT> & msg)
   {
-    for (size_t i = 0; i < Size; ++i) {
-      if (pool_[i].msg_ptr_ == msg) {
-        pool_[i].used = false;
-        return;
+    (void)msg;
+
+    // What we really want to do here is to figure out whether the user has taken an additional
+    // reference to the message, and only add it to the free list if that is *not* the case.
+    // However, we can't really do that for the currently passed-in msg; it can have an arbitrary
+    // reference count due to the mechanisms of rclcpp.  Instead, we look at all the rest of the
+    // pointers, and add the ones that the user has released into the free pool.
+    // We do the same thing in borrow_message(), so if the user has a pool of size 1
+    // (or only one free slot), we'll always find it.
+
+    std::lock_guard<std::mutex> lock(pool_mutex_);
+    if (free_list_.size() == 0) {
+      for (size_t i = 0; i < Size; ++i) {
+        if (pool_[i].use_count() == 1) {
+          free_list_.push_back(i);
+        }
       }
     }
-    throw std::runtime_error("Unrecognized message ptr in return_message.");
   }
 
 protected:
-  struct PoolMember
+  template<size_t N>
+  class CyclicSizeTArray
```

Comment:
nitpick: is this just a circular buffer? Seems like we could call it that, or something like `CircularBuffer<T, N>`? "Cyclic<T>Array" feels a bit like trying to mention a character from another franchise without soliciting a copy-right strike, lol. I don't feel strongly about this, just struck me as weird.

Reply:
> nitpick: is this just a circular buffer? Seems like we could call it that, or something like `CircularBuffer<T, N>`? "CyclicArray"

Yeah, it's a circular buffer.  The name `CyclicArray` came about because I was trying to show that it had no allocations at runtime, like `std::array`.  Maybe `CircularArray` instead?  I don't really have a strong opinion.

Reply:
out of curiosity, can we just use sized `std::queue` thing? it would be easier to implement this class?

Reply:
I'm not sure @fujitatomoya, maybe? I'd have to look at how the functions in `std::queue` work. My concern was mostly about naming tbh. It does feel like it could be generalized, e.g. not specific to `size_t`.